### ISSUE TITLE
Add alternative to force push for updating Renovate branches

### ISSUE THREAD

**Issue by jparnone** [renovatebot/renovate#8239]
**What would you like Renovate to be able to do?**

In instances where force push is disabled by git administrators (typically enterprises), the Renovate process errors when attempting to update branches. If an alternative to force push is in place as an opt-in config that can get user around this challenge.

**Did you already have any implementation ideas?**
Initial thought was to create a new config option, gitDeleteBeforePush, that can be enabled to switch from force push to delete branch and push branch. Initial PR impl: https://github.com/renovatebot/renovate/pull/8233

Initial comments for discussion:

> With this change you won't get any further updates after the pr is closed. So you will get a lot of noise because of renovate ignore pr comments. An you need to reopen the pr manually.

Agree on the noise but this would be a trade off some may be willing to take. For manual reopen that you mentioned, would that be related to how Renovate will ignore PRs that are closed? From my small test on git hub, it seems to have been OK since the initial PR was updated to include "autoclosed" in the title. Does Renovate update the autoclosed PR for every platform?

>It makes no sense to block force push for branches when you can delete and recreate them. 

Do not disagree but there are instances where this set up is in place with limited ability to change.

>Which particular platform is this for? 

My personal need would be Bitbucket server but would think this would be applicable for all supported platforms.

>And how does it behave when the branch is deleted? Most of them close the associated MR/PR, so I don't think this code alone is enough. 

From my limited testing the branch delete closes the open PR and the same PR was able to be recreated in a future Renovate run (same in the sense to the same final version, but different initial versions i.e. PR1 16.3.0 ->17.0.1, PR2 16.4.0 -> 17.0.1). I was not able to test a conflict scenario. What additional logic do you think would be required, or might there be another option other than delete/push?

**Comment by viceice** [comment]
With delete-push option the following happens:

- first renovate create a branch and pr
- another run requires branch update (stale, conflict, ... ). So branch will be deleted and recreated. 
- Platform autoclose pr on branch delete.
- next renovate run creates ignore comment because of closed pr. Branch / pr will never get updated again

So to unblock the update you:
- need to reopen pr
- or rename pr, so renovate can create a new pr

On both ways the pr will be closed and blocked on every renovate branch update.

**Comment by jparnone** [comment]
Created https://github.com/jparnone/renovate/tree/feature/8239-make-force-push-optional to update preferred config option name from initial PR.

**Comment by jparnone** [comment]
> With delete-push option the following happens:
> 
> * first renovate create a branch and pr
> * another run requires branch update (stale, conflict, ... ). So branch will be deleted and recreated.
> * Platform autoclose pr on branch delete.
> * next renovate run creates ignore comment because of closed pr. Branch / pr will never get updated again
> 
> So to unblock the update you:
> 
> * need to reopen pr
> * or rename pr, so renovate can create a new pr
> 
> On both ways the pr will be closed and blocked on every renovate branch update.

Correct me if I'm following correctly. If we went with option two, prior to deleting the branch, if we checked for an open PR created by Renovate, renamed it to something like "<original PR title> - closed for branch update", then perform the branch delete and push that would allow Renovate to open a new PR under the previous title BAU?

**Comment by viceice** [comment]
Yes

**Comment by rarkins** [comment]
You could *maybe* add a function to the platform layer called something like `invalidatePr()` which (a) renames it, (b) closes it, and (c) removes it from the PR cache. If this function were called prior to deleting and recreating the branch then Renovate could update in the same run.

**Comment by jparnone** [comment]
For platforms like Bitbucket where you can truly delete a PR vs decline or merge only, might it be cleaner to delete the PR, delete the existing branch, and then proceed with the new branch create and push? Then for platforms like Github where the PR can't be deleted, we can use the  ```invalidatePr()``` suggestion above.

**Comment by jparnone** [comment]
Rarkins I followed your outline for the invalidatePr idea and pushed a working copy to my fork: https://github.com/renovatebot/renovate/compare/master...jparnone:feature/8239-make-force-push-optional

I created an optional platform method for ```invalidatePr()```, implementing at the moment only for my immediate need in bitbucket-server. I'm checking if the method exists on the platform and if the ```gitDeleteBeforePush``` flag is set and then following the invalidate PR, delete branch, push branch path.

For my tests in a local bitbucket container with force push disabled, I would:

1. Run Renovate once to create a PR
2. Manually up the version slightly to trigger a conflict
3. Rerun Renovate

In these tests, previous PRs are declined with the title updated to add '- autoclosed' to the end and new PRs matching the original title and target versions are created. I'm still considering a delete of the original PR since Bitbucket supports it and it would avoid some unnecessary clutter and may cut down on potential extra notifications.

Do you think there might be any other gotchyas? I have not yet tested with any other configs such as ```recreateClosed``` but I would think the PR rename and cache update should take care off any potential issues.

**Comment by rarkins** [comment]
For your purposes, I agree that it would make most sense for `invalidatePr()` to be deleted instead of renamed/closed.

`gitDeleteBeforePush` should be a platform option which is passed into `initPlatform()` and initialized then (instead of passed per-call in `commitFiles`).

**Comment by jparnone** [comment]
Thanks for that feedback, that's a good idea. I'll incorporate that update and continue working on tests.

**Comment by jparnone** [comment]
When switching to delete vs rename for the PR, it aborts during the first run in the branch worker because of the below refresh call. When it tries to refresh the PR (now deleted at this point in execution) it will return a 404 and trigger the catch. I'm hesitant to start modifying this too much for this single use case. 

The rename flow works fine without additional work needed and while does leave baggage behind which I'd rather not do, it would make a consistent behavior for every platform. What are your thoughts?

```
if (branchPr && platform.refreshPr) {
  await platform.refreshPr(branchPr.number);
}
```

I also updated passing the new config from in each commitFile. I basically expose 2 setters in the GIT util that I call from ```initPlatform()```. I couldn't see another existing pattern to reference the platform values within the util.

If you think moving these discussions into a PR would be better at this point let me know and I can put one up.

**Comment by rarkins** [comment]
@viceice any ideas re: the refreshPr problem? Both these are Bitbucket only IIRC

**Comment by viceice** [comment]
I don't like any of those solutions, as it's a bitbucket server config issue.

To help @jparnone I'm ok with easiest solution. But that needs to be hidden with a feature flag.

BTW: All my bitbucket server will be removed at end of Q1/2021 (because of the licensing changes), so I'm no longer able to test on that platform.

**Comment by jparnone** [comment]
@rarkins do you have any other feedback for this issue before reviewing the PR? I did initially take a look at potential other options but this still seemed the best ultimately with the lowest impact. The main downside is some more notification noise than normal but mailbox rules make quick work of it. Per @viceice this will only apply when the flag is on and only applies to Bitbucket-server based on my and others needs for this need on this platform.

**Comment by rarkins** [comment]
@jparnone sorry, can you summarize now what you recommend we do? I read above and not sure we agreed on a conclusion

**Comment by jparnone** [comment]
Np @rarkins. 

The PR up is focused just on bitbucket-server since that's my immediate need and it was the same platform for another closed issue I saw which had the same request. 

This optional flow to avoid force push is hidden by a new config, gitDeleteBeforePush (defaults to false), as @viceice mentioned he would want. 

In runs for bitbucket-server with the flag set to true, any updates to an existing branch will do the following:

1- Rename an existing PR by appending " - autoclosed" to the current title
2- Remove PR from internal cache
3- Delete the remote branch (which will cause the existing PR to be declined)
4- Push the new branch. 

Note any other platform runs with the flag set to true will not perform differently and continue to use force push for updates. 

Ideally the PR would be deleted vs declined, but doing so would require more changes since Renovate will see the repo has changed and stop the run. Based on the comments in this thread and from what I could see, I don't think the changes required for this vs the value it creates is worth it. @viceice from what I could tell was agreeable to this approach since it was a simpler way.

**Comment by rarkins** [comment]
Is step 5 "Create a new PR" ?

**Comment by jparnone** [comment]
Yes it is. I didn't mention that specifically because that's then back in the normal Renovate flow.

**Comment by rarkins** [comment]
So your idea scenario would be that the previous PR is completely deleted and a new one created? And yes it would be best if it follows the normal flow for PR creation limits.

**Comment by jparnone** [comment]
It would be a little cleaner to delete vs leave the PR in decline but in practice wouldn't make a difference to the SDLC. If this was added in the future to GitHub this would be the only option since PRs cannot be deleted from my understanding. Notification noise wise it would be similar as well.

**Comment by rarkins** [comment]
Today do you minimize rebasing as much as possible, e.g. rebase only when conflicted, or a newer version is available?

**Comment by jparnone** [comment]
Today I personally don't use rebase very often other than a local fetch and rebase vs a merge. Though that's more out of happenstance than a conscience effort.

**Comment by jparnone** [comment]
Just noticed my PR went into conflict so I updated it just now so it should be good to go. 

@rarkins Any other questions before the PR review?

**Comment by rarkins** [comment]
As mentioned in the PR, it's adding a lot of code that complicates our future maintenance.

Let's step back and consider the problem at a higher level.

The Bitbucket Server is configured to disable force pushes to any branch. Is that a global setting or per-repo? Out of interest, can it be detected via API?

Why does Renovate force push to branches?
1. rebaseWhen=behind-base branch: if it's behind the base branch and users want to keep the branch up to date. Should not be configured here
2. rebaseWhen=conflicted: Renovate doesn't have the capabilities to resolve conflicted branches on its own the git way. Force pushing or creating a new branch altogether is really the only solution
3. A new version is released (e.g. current version is 1.0.0, a PR exists for 1.0.1, and then 1.0.2 becomes available)

For case (3), we don't *need* to force push. We could in theory add a new commit on top. The problem is that our autoreplace works by finding references to the *current* value and replacing it with the new value. But in that branch, the current value would have already been replaced (e.g. 1.0.0 with 1.0.1). So this itself would add a lot of complexity to handle, and potentially nasty edge cases.

What if Renovate simply never updates the branches? This would leave two scenarios:
- A conflicted branch, but the version is right, or
- A branch with a not-latest version

For conflicted, the user would need to either resolve the conflict or delete the PR + branch. For non-latest, they'd need to delete it or merge it to make way for the next update. @jparnone is this how you are operating today, or is there some other problem I haven't thought about?

Something else: it was noted that if Renovate simply deletes the branch then Bitbucket will close the PR. By default, that closed PR would *block* future Renovate PRs, but we do have an option to recreateClosed=true. Enabling that would mean you couldn't ignore PRs by closing them intentionally.

**Comment by viceice** [comment]
Maybe it would help to have an option to always separate all updates (event minor and patch), so he can simply close unwanted pr's.

so with `rebaseWhen=never` he gets a solution without force rebase.

**Comment by rarkins** [comment]
Yeah, it would need to be like `separateAllUpdates` and has one branch per dependency version. Although it sounds bad, it would probably end up with less PRs that the current idea.

**Comment by jparnone** [comment]
Thanks for the feedback. I may have confused the earlier suggestions made as the updates that should be made vs initial concepts ideas.

Documentation and test updates aside, are the concerns around the maintainability on the new invalidatePr method in the Bitbucket platform code or the conditional added for determining to delete and push vs force push? Similarly is there a concern on the approach generically or more so on the implementation of the approach?

> The Bitbucket Server is configured to disable force pushes to any branch. Is that a global setting or per-repo? Out of interest, can it be detected via API?

I'm trying to solve it a global level. Bitbucket's API does have a way to retrieve a list of hooks and their status but APIs can be restricted on an individual basis as well so this may or may not be available.


> For conflicted, the user would need to either resolve the conflict or delete the PR + branch. For non-latest, they'd need to delete it or merge it to make way for the next update. @jparnone is this how you are operating today, or is there some other problem I haven't thought about?

No niche cases, your statements are correct. Any challenge today would be manually resolved to allow Renovate to succeed on subsequent runs which works, but would be a pain for the end user to deal with.
 

> Something else: it was noted that if Renovate simply deletes the branch then Bitbucket will close the PR. By default, that closed PR would block future Renovate PRs, but we do have an option to recreateClosed=true. Enabling that would mean you couldn't ignore PRs by closing them intentionally.

For my proposed implementation, before deleting and repushing the branch, I get the open PR and update the name to append "- autoclosed" to it. When the branch is deleted it's this PR with the updated name that is then declined. Doing so allows Renovate to recreate the PR after the new branch is created and pushed within a single run since Renovate would be looking for a declined PR named something like "My Pr Title" vs "My Pr Title - autoclosed". 

I have not tested my code specifically with the recreateClosed option but would think there would be no issue based on the way Renovate, from my understanding, is keying off of the name and since the rename and recreation occurs within the same run. Is there more to the recreateClosed option I might have missed?

**Comment by rarkins** [comment]
> Documentation and test updates aside, are the concerns around the maintainability on the new invalidatePr method in the Bitbucket platform code or the conditional added for determining to delete and push vs force push? Similarly is there a concern on the approach generically or more so on the implementation of the approach?

It's a little bit of everything. Just in general the number of lines that need to be added and maintained indefinitely is very high, and it's not particularly "clean". However I can't think of any cleaner way to achieve the same functionality, so the reality seems to be that we can't do this without it messing up the code base.

Here's another idea: make the branch be based off the exact `toVersion` in the update. This way if you have an update to e.g. `express@4.16.0` then the branch would be named like `renovate/express-4.16.0`. If `express@4.17.0` is released before you merge `4.16.0` then you'd get a new branch/PR created using `renovate/express-4.17.0` and the existing one would be autoclosed *using existing logic*.

That then leaves the case of conflicted PRs. You'd have to either (a) manually de-conflict, or (b) rename and close manually.

The above could probably be done today using configuration-only. Are you interested in testing it out?

**Comment by MartinodF** [comment]
We (Klarna, ~1600 engineers) are in the same situation: our Bitbucket server has force push disabled globally for compliance and auditing purposes, and that's not really something we can change. We're testing out renovate on a small subset of repositories, but this issue prevents us from doing a wider rollout.

We tried deleting all renovate branches and open PRs (thus also avoiding having to set `recreateClosed`) before running and it works, but it needs to be handled by renovate itself only on branches that need it with something like #8398 to be a viable choice.

Would you be open to reconsidering this?

**Comment by rarkins** [comment]
I'm open to reconsidering it (that's why we kept the issue open), but it's going to require quite a bit of planning/research to think about a viable approach. It is definitely not simple. If you or any of your colleagues can dedicate some time to helping us scope out the work and and how the new git logic would work, then we can maybe work out a viable solution. Let me know.

Currently when rebasing:
- Start off the latest main branch
- Apply latest changes
- Force push single commit

New Scenario (an idea):
- Revert previous commit?
- Merge from main branch (should have no merge conflicts..)
- Apply latest changes
- Push all commits on top of existing branch

Without the revert, conflicted branches would be very hard to resolve automatically.

**Comment by MartinodF** [comment]
Hi @rarkins, thank you for getting back to me so quickly!
Yes, I'm more than willing to dedicate as much time as needed to find a solution to this issue 😃

I guess revert + merge + new commit should work in theory, though I'm slightly afraid there's some edge case we're not considering. Is it always a single commit to revert, or could it be multiple?

Just out of curiosity, why did you exclude the possibility of deleting branches and PRs and then starting over? I was initially thinking something along the lines of `rebaseStrategy: "delete-and-recreate"`.

**Comment by rarkins** [comment]
I thought that deleting and starting over would be *very* noisy and unacceptable to most teams. But it could be feasible technically if you think the noise would not a problem for you?

**Comment by viceice** [comment]
I still think your compliance settings are strange.
You are not allowed to force push a branch, but you can delete and recreate it, which is effectively the same as a `force-push`.

I think the revert commit would be best solution, as renovate will currently always only create one commit per branch.

@rarkins this should definetly be disabled by default and only be configurable globally.

**Comment by MartinodF** [comment]
Yes, I agree the settings are strange and I'm not advocating for them. I would guess there must be some difference in the audit trail which makes it easier to accept deleting the entire branch rather than rewriting it, but I'm not really the right person to answer that question 🤷

The revert commit strategy might also be more useful in general since it's not really platform-specific, though I'm not sure what the code looks like! I also completely agree the setting should be disabled by default and global.

Can I offer you any more information about our situation, or is that enough?

**Comment by rarkins** [comment]
I think this is ready for someone to experiment with, but I consider it low priority so it's unlikely to be someone from the core team unless another need arises.

I think add a new globalOnly config option `rebaseMode` which defaults to `force-push` but can be changed to `revert`.

And in such cases it needs to try what I described above:
- Revert the previous commit (assuming it's not edited.. in which we abort and it needs manual intervention)
- Merge commit from base branch
- Apply new commit on top
- Push changes to origin (non-force)

**Comment by mk185147** [comment]
There is another use case when such option would help. We use [GitVersion](https://gitversion.net/) for versioning our artefacts which doesn't work with force push as a new build can have same semantic version which fails violates the requirement for the versioning.

Both options creating new PR and revert + new commit would work as a solution.

**Comment by bartekbluestone** [comment]
is there any change for adding such functionality? For now our company bitbucket for security reason disabled pushing with force, and we cannot disable it for selected users.
Is there any other way to achieve that?

**Comment by sigma67** [comment]
Is there any other option to make renovate work for repositories that prohibit force pushing to all branches?
